edb/server/net_worker.py

```diff
@@ -0,0 +1,243 @@
+                if tenant.accept_new_tasks
+            ]
+            if tasks:
+                await asyncio.wait(tasks)
+        except Exception as ex:
+            logger.debug("HTTP worker failed", exc_info=ex)
+        finally:
+            await asyncio.sleep(
+                POLLING_INTERVAL.to_microseconds() / 1_000_000.0
+            )
+
+
+class HTTPMethod(str, Enum):
+    GET = 'GET'
+    POST = 'POST'
+    PUT = 'PUT'
+    DELETE = 'DELETE'
+    HEAD = 'HEAD'
+    OPTIONS = 'OPTIONS'
+    PATCH = 'PATCH'
+
+
+@dataclass
+class ScheduledRequest:
+    id: str
+    method: HTTPMethod
+    url: str
+    body: typing.Optional[bytes]
+    headers: typing.Optional[dict[str, str]]
+
+    @staticmethod
+    def from_db_record(record: dict) -> ScheduledRequest:
+        id = record['id']
+        if not isinstance(id, str):
+            type_name = type(id).__name__
+            raise TypeError(f"Expected 'id' to be str, got {type_name}")
+
+        method = record['method']
+        if not isinstance(method, str):
+            type_name = type(method).__name__
+            raise TypeError(
+                f"Expected 'method' to be str, got {type_name}"
+            )
+        else:
+            try:
+                method = HTTPMethod(method)
+            except ValueError:
+                raise ValueError(f"Invalid HTTP method: {method}")
+
+        url = record['url']
+        if not isinstance(url, str):
+            type_name = type(url).__name__
+            raise TypeError(
+                f"Expected 'url' to be str, got {type_name}"
+            )
+
+        body = record.get('body')
+        if body is not None and not isinstance(body, bytes):
+            type_name = type(body).__name__
+            raise TypeError(
+                f"Expected 'body' to be bytes or None, got {type_name}"
+            )
+
+        maybe_headers = record.get('headers')
+        if maybe_headers is not None:
+            if not isinstance(maybe_headers, list):
+                type_name = type(maybe_headers).__name__
+                raise TypeError(
+                    f"Expected 'headers' to be list or None, got {type_name}"
+                )
+            if not all(
+                isinstance(name, str) and isinstance(value, str)
+                for name, value in maybe_headers
+            ):
+                raise TypeError(
+                    "Expected 'headers' to be a list of tuples of str"
+                )
+            else:
+                maybe_headers = typing.cast(
+                    list[tuple[str, str]], maybe_headers
+                )
+        headers = dict(maybe_headers) if maybe_headers else None
+
+        return ScheduledRequest(
+            id=id,
+            method=method,
+            url=url,
+            body=body,
+            headers=headers,
+        )
+
+
+async def http_send(request: ScheduledRequest) -> httpx.Response:
+    async with httpx.AsyncClient() as client:
```

Comment:
Maybe we can cache an `httpx.AsyncClient` per tenant to pool connections, but that also leads to racing cookie issues among concurrent requests. So yeah, this is probably fine.

Reply:
Is the idea of caching AsyncClient to minimize number of Python object invocations? Is it expensive to instantiate? Or would it reuse some underlying resources?

Reply:
> reuse some underlying resources

Yes, it reuses the TCP/HTTP connections with `keep-alive`, avoiding some connection establishment time (which we don't care much, because it is supposed to be async and slow). In case of HTTP/2, it could multiplex concurrent requests over the same TCP connection, allowing higher concurrency with minimal resource.

The cost to create an `AsyncClient` is not that much, Python objects are also cleared away after `__aexit__()`. However, the cookies store is shared among all requests on the same client, this may or may not be the expected behavior. For example, `AsyncClient` would follow redirections; if the redirected request depends on the cookie from the initial response, 2 concurrent requests with both redirections on the same domain may step on each other's feet.

In the end of the day, it's still simpler to just use a separate client per request. But I guess in the future we can make it an option/config for the user to decide if needed.

Reply:
It's handy that the client can handle cookies, but I don't think any user (ourselves included) should depend on something stateful like cookies when using this module. Maybe there is a way to disable cookies altogether?

Reply:
Oh, one other nice thing we'd get out of sharing a client is that we can set the max connections in the client itself to avoid managing the concurrency directly ourselves.

Reply:
> Maybe there is a way to disable cookies altogether?

https://github.com/encode/httpx/issues/2992 looks like we'll need to hack our own way in.

> we can set the max connections in the client itself

Yep, there's also that. And it'll be the hardware max connections in case of HTTP/2 (like the number of FDs I believe), which is something we really want I guess (instead of logical number of concurrent requests).

Reply:
I think it makes sense to have a null cookie jar, share a single AsyncClient per-tenant, and set the max connections on the client. I'll throw that together and we can iterate from there if we want something more robust.

Reply:
@fantix ⚡ [`b7f7731` (#7742)](https://github.com/edgedb/edgedb/pull/7742/commits/b7f773114d32e6e97ae936dc832c5e6c29c4bd4e)